### ISSUE TITLE
TensorFlow master branch (tf-nightly) compatibility

### ISSUE THREAD

**Issue by tgaddair** [horovod/horovod#1688]
See: https://github.com/tensorflow/tensorflow/issues/35138

In TensorFlow 2.1, `experimental_run_tf_function` was removed from `tf.keras.Model.compile` at this commit: c73c99c#diff-de9b96ac2d81503324cbbbe21732031fR1159

In Horovod, this flag / graph mode is necessary in order for `Optimizer.get_gradients()` to be called, which aggregates gradients across workers. Since this flag has been removed, distributed training in Horovod with tf.keras isn't working in the example scripts for TensorFlow 2.1.

@abditag2 is currently investigating a workaround.  Note that we cannot perform the allreduce aggregation in apply_gradients due to interactions with gradient clipping and loss scaling (see horovod/horovod#1347).

**Comment by nvcastet** [comment]
To be clear, TF 2.1.0 release works fine because the https://github.com/tensorflow/tensorflow/commit/c73c99ca3e0bacf2bca313f270bb3eae28869530 commit is not included in it. It is currently in `master`, so the next release of TF will have the issue.

**Comment by tgaddair** [comment]
Thanks @nvcastet, good catch.  I'll update the issue.

**Comment by abditag2** [comment]
we have a temporary fix for this in https://github.com/horovod/horovod/pull/1692

However, we are working with TF team to come up with a better solution here: https://github.com/tensorflow/tensorflow/issues/36398

**Comment by Leslie-Fang** [comment]
I will meet the same error when use TF2.1 and horovod 0.19.0

**Comment by tgaddair** [comment]
@Leslie-Fang, are you setting `experimental_run_tf_function=False` in your script?

**Comment by Leslie-Fang** [comment]
@tgaddair  Yes, I set `experimental_run_tf_function=False`.
Downgrade the TF t0 2.0 fix this problem.
